ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -510,6 +502,14 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = do
 
     processLoE curChain (LgrDB.ledgerDbCurrent ledgerDB) (writeTVar cdbLoEFrag) cdbLoE
 
+    loeFrag <- atomically (cross curChain <$> readTVar cdbLoEFrag) <&> \case
+        Just (_, frag) -> frag
+        -- We don't crash if the LoE fragment doesn't intersect with the selection
+        -- because we update the selection _after_ updating the LoE fragment, which
+        -- means it could move to another fork or beyond the end of the LF, depending
+        -- on the implementation of @updateLoEFrag@.
+        Nothing        -> AF.Empty (AF.anchor curChain)
```

Comment:
`pointHash tipPoint == headerPrevHash hdr` in the checks below is not a sufficient condition to extend the selection. In addition, it should be checked that the new header is a descendent of the tip of the LoE fragment, or it is contained in the LoE fragment. 

Otherwise, if the block belongs to an adversary that has been disconnected, the node will continue to extend the selection on a chain which is served by the adversary.

This raises the question that LoE should determine on which chain the selection is supposed to be, unlike the current code that resets LoE if it is on a different chain.

Reply:
> `pointHash tipPoint == headerPrevHash hdr` in the checks below is not a sufficient condition to extend the selection. In addition, it should be checked that the new header is a descendent of the tip of the LoE fragment, or it is contained in the LoE fragment.
> 
> Otherwise, if the block belongs to an adversary that has been disconnected, the node will continue to extend the selection on a chain which is served by the adversary.

It should be unproblematic/even expected if we are selecting some adversarial blocks after the adversary has been disconnected (eg because the adversary sent them very quickly, and they were still in the ChainSel queue after we disconnected from the adversary, waiting to be processed) as long as we don't select more than `k` blocks past the LoE. `computeLoEMaxExtra` will return `Nothing` if that is not the case, and hence we won't adopt the block.

Reply:
I think you are correct in essence. I have made `computeLoEMaxExtra` more precise in 716d975.